Doc: Custom templates docs (#3150)
aristanetworks/ansible-avd: ansible_collections/arista/avd/roles/eos_cli_config_gen/docs/how-to/custom-templates.md

<!--
  ~ Copyright (c) 2023 Arista Networks, Inc.
  ~ Use of this source code is governed by the Apache License 2.0
  ~ that can be found in the LICENSE file.
  -->

# Custom Templates

With Custom Templates for the `eos_cli_config_gen` role, the user can create additional CLI configuration for features that are not covered by the built-in templates.

## Create a Custom Template

Custom templates should be written in Jinja2.

For details on how to create Jinja2 templates, see [Templating Jinja2](https://docs.ansible.com/ansible/latest/playbook_guide/playbooks_templating.html#templating-jinja2)

> Custom templates can utilize any AVD variable for variable substitution including additional structured configuration variables created by a custom template in the `eos_designs` role. For details on how to create additional structured configuration variables see [Custom Templates for eos_designs](../../../eos_designs/docs/how-to/custom-templates.md) 

Reviewer: The simpler way to provide variables is to use `custom_structured_configuration_` or the various `structured_config` keys in the data models. There is another howto on that, and the vars will show up in the output of `eos_designs` so they can indeed be picked up by the custom template in `eos_cli_config_gen`.

Author: Amended in the latest push.

> 

Reviewer: Maybe add similar details here on where the file should be placed, for Jinja2 to be able to pick it up. (Hint: It only works below the playbook dir, so you might have to use symlinks if it is placed elsewhere)

Author: Amended in the latest push.

> ## Adding the Custom Template to groupvars

Reviewer: ```suggestion
## Adding the Custom Template to group vars
```

Author: Fixed in the latest push.

> For the custom template to be automatically discovered and rendered by `eos_cli_config_gen`, a variable that references the template should be added to `groupvars`.

Reviewer: ```suggestion
For the custom template to be automatically discovered and rendered by `eos_cli_config_gen`, a variable that references the template should be added to the inventory group vars.
```

Author: Fixed in the latest push.

<!--
  ~ Copyright (c) 2023 Arista Networks, Inc.
  ~ Use of this source code is governed by the Apache License 2.0
  ~ that can be found in the LICENSE file.
  -->

# Custom Templates

> With Custom Templates for the `eos_cli_config_gen` role, the user can create additional CLI configuration for features that are not covered by the built-in templates.

Reviewer 2: ```suggestion
With Custom Templates for the `eos_cli_config_gen` role, the user can create additional CLI configurations for features not covered by the built-in templates.
```

## Create a Custom Template

Custom templates should be written in Jinja2.

> For details on how to create Jinja2 templates, see [Templating Jinja2](https://docs.ansible.com/ansible/latest/playbook_guide/playbooks_templating.html#templating-jinja2)

Reviewer 2: ```suggestion
For details on how to create Jinja2 templates, see [Templating Jinja2](https://docs.ansible.com/ansible/latest/playbook_guide/playbooks_templating.html#templating-jinja2).
```

> Custom templates can utilize variables defined by all the AVD variable definition mechanisms. For more details see the explanation in [Custom Structured Configuration](../../../eos_designs/docs/how-to/custom-structured-configuration.md)

Reviewer 2: ```suggestion
Custom templates can utilize variables defined by all the AVD variable definition mechanisms. For more details, see the explanation in [Custom Structured Configuration](../../../eos_designs/docs/how-to/custom-structured-configuration.md).
```

> If the required variable is not available by any of these mechanisms, it can be defined using a custom template in the `eos_designs` role. For more details see [Custom Templates for eos_designs](../../../eos_designs/docs/how-to/custom-templates.md)

Reviewer 2: ```suggestion
Suppose the required variable is not available by any of these mechanisms. In that case, it can be defined using a custom template in the `eos_designs` role. For more details see [Custom Templates for eos_designs](../../../eos_designs/docs/how-to/custom-templates.md).
```

At runtime, `eos_cli_config_gen` searches for custom templates in the following location, therefore, custom templates should be saved here:

```text
<path to users AVD implementation>/playbooks/templates/<template name>
```

> The "templates" folder does not exist by default so will need to be created.

Reviewer 2: ```suggestion
The "templates" folder does not exist by default, so it must be created.
```

## Adding the Custom Template to group vars

For the custom template to be automatically discovered and rendered by `eos_cli_config_gen`, a variable that references the template should be added to the inventory group vars.

The format of the variable should be as follows:

```yaml
custom_templates:
  - custom_template_name.j2
```

For more details, see [Extensibility with Custom Templates](https://avd.sh/en/stable/roles/eos_cli_config_gen/docs/role-configuration.html#extensibility-with-custom-templates)

## Adding the Custom Template to the `eos_cli_config_gen` Role

> This happens automatically as part of the `Generate eos intended configuration` task in `~/.ansible/collections/ansible_collections/arista/avd/roles/eos_cli_config_gen/tasks/main.yaml`.
> 
> <!-- ![Figure 1: /eos_cli_config_gen/templates/eos-intended-config.j2](../../../media/main_yml.png) -->
> 
> <div style="text-align:left; width:500px">
>   <img src="../../../../media/main_yml.png" alt="screen shot of main.yml file" />
> </div>
> 
> This task renders the main `eos-intended-config.j2` template which has multiple include statements. One of these include statements pertains to custom templates and causes any custom templates to be included in the rendering process.
> 
> <!-- ![Figure 2: /eos_cli_config_gen/templates/eos-intended-config.j2](../../../media/eos-intended-config.png) -->
> 
> <div style="text-align:left; width:400px">
>   <img src="../../../../media/eos-intended-config.png" alt="screen shot of snippet of main template" />
> </div>

Reviewer: ```suggestion
Custom templates are rendered automatically when the `custom_templates` variable is set. The custom templates are always rendered _after_ the builtin `eos_cli_config_gen` templates.
```
